### PR TITLE
Feature: Town name filtering

### PR DESCRIPTION
**Pull request by telk5093**
This is a re-PR of #6983, which implements the filtering feature for town name list like below:
(I reopen a PR since I have no idea how to change branch of previous PR)
image

Now it does not blink and seems stable unlike #6983 as @LordAro said.
Nevertheless, please let me know if there any problem with my commit and also please aware that I'm very new to both C/C++ and OpenTTD's codes.

### CODE REVIEW

File: src/town_gui.cpp [hunk 1]
```diff
@@ -662,6 +663,10 @@ struct TownDirectoryWindow : public Window {
 	static const StringID sorter_names[];
 	static GUITownList::SortFunction * const sorter_funcs[];
 
+	StringFilter string_filter;             ///< Filter for towns
+	QueryString townname_editbox;           ///< Filter editbox
+	std::vector<bool> town_items_shown; ///< Map of the filtered out town names
```
Comment (LordAro) [inline src/town_gui.cpp]: this is unused

Comment (telk5093) [reply src/town_gui.cpp]: Thanks for notice, fixed.

**Comment by nielsmh** [comment]
Remember to run the `generate_widget` and `squirrel_export` scripts after adding a new widget ID, and commit the updated script API files.

File: src/town_gui.cpp [hunk 2]
```diff
@@ -689,8 +693,8 @@ struct TownDirectoryWindow : public Window {
 	/** Sort by town name */
 	static bool TownNameSorter(const Town * const &a, const Town * const &b)
 	{
-		static char buf_cache[64];
-		char buf[64];
+		static char buf_cache[MAX_LENGTH_TOWN_NAME_CHARS];
+		char buf[MAX_LENGTH_TOWN_NAME_CHARS];
```
Comment (LordAro) [inline src/town_gui.cpp]: i feel like these should be `MAX_LENGTH_TOWN_NAME_CHARS * MAX_CHAR_LENGTH`, same as the QueryString - a townname could conceivably be constructed out of 64 4-byte characters, and of course "char" is just a byte

Ideally the 64 -> constant changes should be in a separate commit (there may be other instances of this being incorrect as well)

Comment (telk5093) [reply src/town_gui.cpp]: I changed them and seperated commit as you said, thanks a lot.

**Comment by telk5093** [comment]
@nielsmh I had no idea I should run ``squirrel_export`` additionally. Thanks for noticing me and has done.

**Review by LordAro — APPROVED** [review]
